src/radchallenge.c

```diff
@@ -0,0 +1,210 @@
+    /* If the IP Address is provided via Command-line, take it for processing. Else,
+     * use localhost as default.
+     */
+    if(server_ip == NULL)
+        server_ip = SERVER_ADDR;
+
+    snprintf(auth_server_ip, sizeof(auth_server_ip), "%s%s", server_ip,
+             AUTH_PORT_PASSWORD);
+
+    if (rc_add_config(rh, "authserver", auth_server_ip, "config", 0) != 0)
+    {
+        printf("ERROR: Unable to set authserver.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+
+    /* Done setting configuration items */
+
+    /* Make dictionary */
+
+    if (rc_dict_addattr(rh, "User-Name", 1, PW_TYPE_STRING, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute User-Name.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "Password", 2, PW_TYPE_STRING, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute Password.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "NAS-IP-Address", 4, PW_TYPE_IPADDR, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute NAS-IP-Address.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "NAS-Port-Id", 5, PW_TYPE_INTEGER, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute NAS-Port-Id.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "Service-Type", 6, PW_TYPE_INTEGER, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute Service-Type.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "Reply-Message", 18, PW_TYPE_STRING, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute Password.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+    if (rc_dict_addattr(rh, "State", 24, PW_TYPE_STRING, 0) == NULL)
+    {
+        printf("ERROR: Can not add attribute Password.\n");
+        rc_destroy(rh);
+        exit(1);
+    }
+
+    /* Initialize socket related info in RADIUS Handle */
+    if (rc_apply_config(rh) == -1)
+    {
+        printf("ERROR: Failed to update Radius handle socket info");
+        exit(1);
+    }
+
+    if (rc_avpair_add(rh, &send, PW_USER_NAME, username, -1, 0) == NULL)
+    {
+        printf("ERROR: Failed adding User-Name: to %s\n", username);
+        exit(1);
+    }
+    if (rc_avpair_add(rh, &send, PW_USER_PASSWORD, password, -1, 0) == NULL)
+    {
+        printf("ERROR: Failed adding Password: to %s\n", password);
+        exit(1);
+    }
+    if (use_state)
+    {
+        if (rc_avpair_add(rh, &send, PW_STATE, state, -1, 0) == NULL) {
+            printf("ERROR: Failed adding State: to %s\n", state);
+            exit(1);
+        }
+    }
+    uint32_t service = PW_AUTHENTICATE_ONLY;
+    if (rc_avpair_add(rh, &send, PW_SERVICE_TYPE, &service, sizeof(service), 0) == NULL)
+    {
+        printf("ERROR: Failed adding Service-Type: to %u\n", service);
+        exit(1);
+    }
+
+    int result = rc_auth(rh, 0, send, &received, nullptr);
```

Comment:
Do we need the real auth to test this functionality? I.e., can we test it without involving a server? That will simplify the running of the test.